File: src/WikiParse/Category.php
<?php

namespace WikiParse\Category;

include_once __DIR__ . '/include_it.php';

/*
Usage:

use function WikiParse\Category\get_categories_reg;

*/

function get_categories_reg($text)
{
    $categories = array();

    // هذا التعبير النمطي يستخدم العودية (?R) للتعامل مع الأقواس المتداخلة بشكل صحيح.
    // (?R) تطابق النمط بأكمله مرة أخرى، مما يسمح بمطابقة هياكل متداخلة مثل [[...[...]...]].
    $pattern = "/\[\[\s*Category\s*\:([^\]\]]+?)\]\]/is";
    // $pattern = "/\[\[\s*Category\s*:(.*?)\]\](?!\])/is";

    preg_match_all($pattern, $text, $matches);

    if (!empty($matches[1])) {
        foreach ($matches[0] as $i => $full_match) {
            $category_content = $matches[1][$i];
            // الآن نقوم بتقسيم المحتوى بناءً على "|" للحصول على اسم التصنيف فقط
            $parts = explode('|', $category_content);
            $category_name = trim(array_shift($parts));

            // لا نزال نستخدم المطابقة الكاملة كقيمة في المصفوفة النهائية
            $categories[$category_name] = $full_match;
        }
    }

    return $categories;
}

File: python/src/Parse/Category.py
"""
Category parser module

Implemented from: src/Parse/Category.php

Usage:
    from src.Parse.Category import get_categories_reg
"""

import re


def get_categories_reg(text: str) -> dict:
    """
    Get categories from text using regex
    
    Matches: src/Parse/Category.php - get_categories_reg()
    
    Args:
        text: Text to parse for categories
        
    Returns:
        Dictionary mapping category names to their full match strings
    """
    categories = {}
    
    # Pattern to match category links
    pattern = r'\[\[\s*Category\s*:([^\]\]]+?)\]\]'
    
    matches = re.findall(pattern, text, re.IGNORECASE | re.DOTALL)
    full_matches = re.findall(r'\[\[\s*Category\s*:[^\]\]]+?\]\]', text, re.IGNORECASE | re.DOTALL)
    
    if matches:
        for i, category_content in enumerate(matches):
            # Split on "|" to get just the category name
            parts = category_content.split('|')
            category_name = parts[0].strip()
            
            # Use full match as value
            if i < len(full_matches):
                categories[category_name] = full_matches[i]
    
    return categories
